go/arrow/array/encoded.go

```diff
@@ -0,0 +1,304 @@
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+	buf.WriteByte('[')
+
+	for i := range r.runEnds {
+		if i != 0 {
+			buf.WriteByte(',')
+		}
+		if err := enc.Encode(r.getOneForMarshal(i)); err != nil {
+			return nil, err
+		}
+	}
+	buf.WriteByte(']')
+	return buf.Bytes(), nil
+}
+
+func arrayRunLengthEncodedEqual(l, r *RunLengthEncoded) bool {
+	// types were already checked before getting here, so we know
+	// the encoded types are equal
+	mr := rle.NewMergedRuns([2]arrow.Array{l, r})
+	for mr.Next() {
+		lIndex := mr.IndexIntoArray(0)
+		rIndex := mr.IndexIntoArray(1)
+		if !SliceEqual(l.values, lIndex, lIndex+1, r.values, rIndex, rIndex+1) {
+			return false
+		}
+	}
+	return true
+}
+
+func arrayRunLengthEncodedApproxEqual(l, r *RunLengthEncoded, opt equalOption) bool {
+	// types were already checked before getting here, so we know
+	// the encoded types are equal
+	mr := rle.NewMergedRuns([2]arrow.Array{l, r})
+	for mr.Next() {
+		lIndex := mr.IndexIntoArray(0)
+		rIndex := mr.IndexIntoArray(1)
+		if !sliceApproxEqual(l.values, lIndex, lIndex+1, r.values, rIndex, rIndex+1, opt) {
+			return false
+		}
+	}
+	return true
+
+}
+
+type RunLengthEncodedBuilder struct {
+	builder
+
+	dt      arrow.DataType
+	runEnds *Int32Builder
+	values  Builder
+}
+
+func NewRunLengthEncodedBuilder(mem memory.Allocator, typ arrow.DataType) *RunLengthEncodedBuilder {
+	return &RunLengthEncodedBuilder{
+		builder: builder{refCount: 1, mem: mem},
+		dt:      arrow.RunLengthEncodedOf(typ),
+		runEnds: NewInt32Builder(mem),
+		values:  NewBuilder(mem, typ),
+	}
+}
+
+func (b *RunLengthEncodedBuilder) Type() arrow.DataType {
+	return b.dt
+}
+
+func (b *RunLengthEncodedBuilder) Release() {
+	debug.Assert(atomic.LoadInt64(&b.refCount) > 0, "too many releases")
+
+	if atomic.AddInt64(&b.refCount, -1) == 0 {
+		b.values.Release()
+		b.runEnds.Release()
+	}
+}
+
+func (b *RunLengthEncodedBuilder) addLength(n uint32) {
+	if b.length+int(n) > math.MaxInt32 {
+		panic(fmt.Errorf("%w: run-length encoded array length must fit in a 32-bit signed integer", arrow.ErrInvalid))
+	}
+
+	b.length += int(n)
+}
+
+func (b *RunLengthEncodedBuilder) finishRun() {
+	if b.length == 0 {
+		return
+	}
+
+	b.runEnds.Append(int32(b.length))
+}
+
+func (b *RunLengthEncodedBuilder) ValueBuilder() Builder { return b.values }
+func (b *RunLengthEncodedBuilder) Append(n uint32) {
+	b.finishRun()
+	b.addLength(n)
+}
+func (b *RunLengthEncodedBuilder) ContinueRun(n uint32) {
+	b.addLength(n)
+}
+func (b *RunLengthEncodedBuilder) AppendNull() {
+	b.finishRun()
+	b.values.AppendNull()
+	b.addLength(1)
+}
+
+func (b *RunLengthEncodedBuilder) NullN() int {
+	return UnknownNullCount
+}
+
+func (b *RunLengthEncodedBuilder) AppendEmptyValue() {
+	b.AppendNull()
+}
+
+func (b *RunLengthEncodedBuilder) Reserve(n int) {
+	b.values.Reserve(n)
+	b.runEnds.Reserve(n)
+}
+
+func (b *RunLengthEncodedBuilder) Resize(n int) {
+	b.values.Resize(n)
+	b.runEnds.Resize(n)
+}
```

Comment:
Do you not need to resize/reserve the validity map?

Reply:
based on the current spec `RunLengthEncoded` arrays have no validity map